internal/cachecontroller/cache_controller.go

```diff
@@ -0,0 +1,160 @@
+
+	"github.com/openfga/openfga/internal/build"
+
+	"go.opentelemetry.io/otel"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/openfga/openfga/pkg/storage"
+)
+
+var (
+	tracer = otel.Tracer("internal/cachecontroller")
+
+	cacheTotalCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: build.ProjectName,
+		Name:      "cachecontroller_cache_total_count",
+		Help:      "The total number of cachecontroller requests.",
+	})
+
+	cacheHitCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: build.ProjectName,
+		Name:      "cachecontroller_cache_hit_count",
+		Help:      "The total number of cache hits from cachecontroller requests.",
+	})
+)
+
+type CacheController interface {
+	DetermineInvalidation(ctx context.Context, storeID string) time.Time
+}
+
+type NoopCacheController struct{}
+
+func (c *NoopCacheController) DetermineInvalidation(_ context.Context, _ string) time.Time {
+	return time.Time{}
+}
+
+func NewNoopCacheController() CacheController {
+	return &NoopCacheController{}
+}
+
+type InMemoryCacheController struct {
+	ds               storage.OpenFGADatastore
+	cache            storage.InMemoryCache[any]
+	ttl              time.Duration
+	iteratorCacheTTL time.Duration
+
+	sf *singleflight.Group
+}
+
+func NewCacheController(ds storage.OpenFGADatastore, cache storage.InMemoryCache[any], ttl time.Duration, iteratorCacheTTL time.Duration) CacheController {
+	c := &InMemoryCacheController{
+		ds:               ds,
+		cache:            cache,
+		ttl:              ttl,
+		iteratorCacheTTL: iteratorCacheTTL,
+		sf:               &singleflight.Group{},
+	}
+
+	return c
+}
+
+func (c *InMemoryCacheController) DetermineInvalidation(
+	ctx context.Context,
+	storeID string,
+) time.Time {
+	ctx, span := tracer.Start(ctx, "cacheController.ResolveCheck")
+	defer span.End()
+	cacheTotalCounter.Inc()
+
+	cacheKey := storage.GetChangelogCacheKey(storeID)
+	cacheResp := c.cache.Get(cacheKey)
+	if cacheResp != nil && !cacheResp.Expired && cacheResp.Value != nil {
+		entry := cacheResp.Value.(*storage.ChangelogCacheEntry)
+		cacheHitCounter.Inc()
+		return entry.LastModified
+	}
+
+	lastModified, err, _ := c.sf.Do(storeID, func() (interface{}, error) {
+		return c.findChangesAndInvalidate(ctx, storeID)
+	})
+
+	if err != nil {
+		return time.Time{}
+	}
+
+	return lastModified.(time.Time)
+}
+
+func (c *InMemoryCacheController) findChanges(ctx context.Context, storeID string) ([]*openfgav1.TupleChange, []byte, error) {
+	opts := storage.ReadChangesOptions{
+		SortDesc: true,
+		Pagination: storage.PaginationOptions{
+			PageSize: storage.DefaultPageSize,
+			From:     "",
+		}}
+	return c.ds.ReadChanges(ctx, storeID, storage.ReadChangesFilter{}, opts)
+}
+
+func (c *InMemoryCacheController) findChangesAndInvalidate(ctx context.Context, storeID string) (time.Time, error) {
+	ctx, span := tracer.Start(ctx, "cacheController.Invalidations")
```

Comment:
could it be useful to annotate this span with the invalidation side effects that occurred?